Using the Debugger (#67)
Ada-Developers-Academy/core-unit-1: intermediate-python-debugging/exercise-using-debugger.md

# Problem Set: Using the Debugger

In this exercise we will use the debugger to identify two errors.  To run the activity clone the code at [AdaGold/python-debugging](https://github.com/AdaGold/python-debugging).  Then set up the project by creating a virtual environment and installing pytest.

> The code contains two bugs, one runtime error crashes the program on execution and the other is a logic error which calculates the incorrect sum for an Order instance.  

Reviewer: This comes across as a little unclear to me.

I think if you convert this to an [Activity or Problem Set template](https://github.com/Ada-Developers-Academy/core/wiki/Templates-and-Stubs), this content would split underneath the different headers in the template easily.

I think you can also cut the part about setting up the project with the venv and installing pytest, and instead in the instructions point them to the repo README and say "Setup the activity by visiting here and following the instructions on the README."

I'd also love an explicit step that says something like "After following the setup instructions on the python-debugging repo, follow the instructions below."

> The code contains two bugs, one runtime error crashes the program on execution and the other is a logic error which calculates the incorrect sum for an Order instance.  

Reviewer: Order instance is the last part and should probably be corrected

> ## Activity

Reviewer: Is this a problem set or activity (check top header and second header and file name)?

In the current curriculum:

Problem Sets are pieces of (ideally) one-night homework. The file should be prefixed `problem-set` (not exercise). Problem sets are something we want to ask students to definitely complete, so usually there's some Learn challenge autograded component to it.

Activities are designed to replace a roundtable, so are ideally written to be one-hour small group activities.

In either case, at some point, it'd be great if it followed a [curriculum template](https://github.com/Ada-Developers-Academy/core/wiki/Templates-and-Stubs_). I recognize that it kinda sucks because you prob didn't know about this when you wrote it, so if you don't convert it to this template, I will eventually :D !

Reviewer: One more note, in my ideal world, the header and file name should match exactly, including the "the"

`activity-using-the-debugger` for "Activity: Using the Debugger"
or `activity-using-debugger` for "Activity: Using Debugger"

> ### 1st Bug

Reviewer: ```suggestion
### First Bug: Debugging with Breakpoints
```

Grammar. Suggesting adding ": Debugging with Breakpoints" to give it more context that helps frame the point of the first bug to students

  File "/Users/username/ada/python/python-debugging/grocery_store/order.py", line 16, in calculate_total
    total += order["products"][i]["price"]
TypeError: unsupported operand type(s) for +=: 'int' and 'str'
```

Hmmmm... It seems that it crashes at line 16 of `grocery_store/order.py`.

Lets put a breakpoint at line 16 of `grocery_store/order.py`.

![Visual of a breakpoint in order.py at line 16](../assets/vs-code-debugger/order-py-breakpoint.png)

Then go to the debugger icon.

![Debugger Icon](../assets/vs-code-debugger/debugger-icon.png)

You **may** need to click on the select list, select `Add Configuration` and choose select `Python: Current File`.

![select debug current python file](../assets/vs-code-debugger/debug-current-python-file.png)

Otherwise you can simpily select "Python: Current File" from the drop down list.  

### Running Debugger

Then you can run the debugger by opening `app.py` and pressing the green triangle "Play" button.

![Debugger Play Button](../assets/vs-code-debugger/start-debugger.png)

You will notice that the application stops in `order.py`.

![Running Debugger](../assets/vs-code-debugger/running-debugger.png)

> ### Finding the bug...

Reviewer: Title case, suggesting removing the `...` and "spoiling" the section by saying the goal is to find the bug with watches

```suggestion
### Finding the Bug with Watches
```

Notice that the loop attempts to add up all the product prices.  In Watch add a watch for `order["products"][i]`

![Adding a watch on a product](../assets/vs-code-debugger/watch-price.png)

<details style="max-width: 700px; margin: auto;">
  <summary>

>     Do you see the problem?

Reviewer: I'd personally prefer a different way to word this. Maybe even just replacing the word "problem" with "unintended behavior" or something similar... but in my opinion, for a details/summary block, the "summary" question should be something that has a clear moment where a student knows they should open this up.

With the current wording, does the student open it up when they see the problem? Or when they don't see the problem and it's been 2 hours? etc lol

  </summary>

>   The price for the products gets set in `product.py`, see if you can find the error, fix it, and then re-run the debugger.  Does it crash now?  If you need add another breakpoint to `product.py` in the `create_product` function and step through the program.

Reviewer: Imo, this content seems more like a mixed set of steps and guidelines (more like board game instructions).

If that's the case, Reformatting it as a list or separate steps might be helpful? (tbh, I'm not sure if that accomplishes what you were intending)

```suggestion
The bug is in `product.py`, and it relates to the price of products.

Hypothesize a cause for the error, make changes to fix it, and then re-run the debugger.

Repeat this process until the program stops crashing, and the bug is fixed. During this debugger process, practice:

- Adding another breakpoint to `product.py` inside the `create_product` function
- Stepping through the program when re-running the debugger
```

> </details>

Reviewer: Truthfully, I don't see a need to put this into a details/summary block... the answer in the details block doesn't seem to be "spoilable"? I don't think it detracts from the student to see as a paragraph underneath (not in a details block).

> ### The 2nd Bug

Reviewer: ```suggestion
### Second Bug: Debugging a Failing Test
```

Removes the "The" to conform to the "First Bug" header, grammar and context

One of the tests is failing, find the failing test in `tests/test_order.py`.  

![Failing test](../assets/vs-code-debugger/exercise-failing-test.png)

Create a breakpoint in the failing test `test_calculate_total_with_multiple_products` and use the step-into and step over commands to trace through the code and see why the test is failing.  You can click on the `Debug Test` link above the test function to run that one test in the debugger.

> 

Reviewer: I like this, but it seems incomplete?
